server/debugsession.go

```diff
@@ -0,0 +1,834 @@
+		}
+	case cadence.Struct:
+		for i, field := range value.StructType.Fields {
+			variable := ds.convertValueToDapVariable(field.Identifier, value.Fields[i])
+			members = append(members, variable)
+		}
+
+	case cadence.Array:
+		for i, element := range value.Values {
+			variable := ds.convertValueToDapVariable(fmt.Sprintf("[%d]", i), element)
+			members = append(members, variable)
+		}
+
+	case cadence.Dictionary:
+		for _, pair := range value.Pairs {
+			variable := ds.convertValueToDapVariable(pair.Key.String(), pair.Value)
+			members = append(members, variable)
+		}
+	}
+	return members
+}
+
+func (ds *debugSession) convertInterpreterValueToDapVariables(
+	inter *interpreter.Interpreter,
+	value interpreter.Value,
+) []dap.Variable {
+	cadenceValue, err := runtime.ExportValue(value, inter, interpreter.EmptyLocationRange)
+	if err != nil {
+		panic(err)
+	}
+	return ds.convertCadenceValueMembersToDapVariables(cadenceValue)
+}
+
+func (ds *debugSession) convertStorageMapToDapVariables(
+	inter *interpreter.Interpreter,
+	value *interpreter.StorageMap,
+) []dap.Variable {
+
+	members := make([]dap.Variable, value.Count())
+
+	it := value.Iterator(inter)
+	for {
+		key, value := it.Next()
+		if key == "" {
+			break
+		}
+
+		cadenceValue, err := runtime.ExportValue(value, inter, interpreter.EmptyLocationRange)
+		if err != nil {
+			panic(err)
+		}
+
+		variable := ds.convertValueToDapVariable(key, cadenceValue)
+		members = append(members, variable)
+	}
+
+	return members
+}
+
+func (ds *debugSession) stackFrames() []dap.StackFrame {
+	invocations := ds.stop.Interpreter.CallStack()
+
+	stackFrames := make([]dap.StackFrame, 0, len(invocations))
+
+	location := ds.stop.Interpreter.Location
+	astRange := ast.NewRangeFromPositioned(nil, ds.stop.Statement)
+
+	startPos := astRange.StartPosition()
+	endPos := astRange.EndPosition(nil)
+
+	locationString := locationPath(location)
+	if location.String() == ds.scriptID {
+		locationString = ds.scriptLocation.String()
+	}
+	stackFrames = append(
+		stackFrames,
+		dap.StackFrame{
+			Source: dap.Source{
+				Path: locationString,
+			},
+			Line:      startPos.Line,
+			Column:    startPos.Column + 1,
+			EndLine:   endPos.Line,
+			EndColumn: endPos.Column + 2,
+		},
+	)
+
+	for i := len(invocations) - 1; i >= 0; i-- {
+		invocation := invocations[i]
+
+		locationRange := invocation.LocationRange
+
+		location := locationRange.Location
+		if location == nil {
+			continue
+		}
+
+		startPos := locationRange.StartPosition()
+		endPos := locationRange.EndPosition(nil)
+
+		locationString := locationPath(location)
+		if location.String() == ds.scriptID {
+			locationString = ds.scriptLocation.String()
+		}
+		stackFrames = append(
+			stackFrames,
+			dap.StackFrame{
+				Source: dap.Source{
+					Path: locationString,
+				},
+				Line:      startPos.Line,
+				Column:    startPos.Column + 1,
+				EndLine:   endPos.Line,
+				EndColumn: endPos.Column + 2,
+			},
+		)
```

Comment:
I feel this could be extracted to a function as it repeats the above logic, and accept range and location and produce stack frame.

Reply:
Good point! 

Refactored in 3a6ae2e638fc6e10f0caba138c7ad76e6232cd20